les/utils/limiter_test.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+const ltTolerance = 0.15
+
+type (
+	ltNode struct {
+		addr, id         int
+		value, exp       float64
+		cost             uint
+		reqRate          float64
+		reqMax, runCount int
+		lastTotalCost    uint
+
+		served, dropped int
+	}
+
+	ltResult struct {
+		node *ltNode
+		ch   chan struct{}
+	}
+
+	limTest struct {
+		limiter            *Limiter
+		results            chan ltResult
+		runCount           int
+		expCost, totalCost uint
+	}
+)
+
+func (lt *limTest) request(n *ltNode) {
+	var (
+		address string
+		id      enode.ID
+	)
+	if n.addr >= 0 {
+		address = string([]byte{byte(n.addr)})
+	} else {
+		var b [32]byte
+		rand.Read(b[:])
+		address = string(b[:])
+	}
+	if n.id >= 0 {
+		id = enode.ID{byte(n.id)}
+	} else {
+		rand.Read(id[:])
+	}
+	lt.runCount++
+	n.runCount++
+	cch := lt.limiter.Add(id, address, n.value, n.cost)
+	go func() {
+		lt.results <- ltResult{n, <-cch}
+	}()
+}
+
+func (lt *limTest) moreRequests(n *ltNode) {
+	maxStart := int(float64(lt.totalCost-n.lastTotalCost) * n.reqRate)
+	if maxStart != 0 {
+		n.lastTotalCost = lt.totalCost
+	}
+	for n.reqMax > n.runCount && maxStart > 0 {
+		lt.request(n)
+		maxStart--
+	}
+}
+
+func (lt *limTest) process() {
+	res := <-lt.results
+	lt.runCount--
+	res.node.runCount--
+	if res.ch != nil {
+		res.node.served++
+		if res.node.exp != 0 {
+			lt.expCost += res.node.cost
+		}
+		lt.totalCost += res.node.cost
+		close(res.ch)
+	} else {
+		res.node.dropped++
+	}
+}
+
+func TestLimiter(t *testing.T) {
```

Comment:
```
    TestLimiter: limiter_test.go:181: Request ratio (0.751915) does not match expected value (0.500000)
    TestLimiter: limiter_test.go:181: Request ratio (0.121453) does not match expected value (0.250000)
    TestLimiter: limiter_test.go:181: Request ratio (0.126632) does not match expected value (0.250000)
    TestLimiter: limiter_test.go:181: Request ratio (0.895618) does not match expected value (0.500000)
    TestLimiter: limiter_test.go:181: Request ratio (0.052434) does not match expected value (0.250000)
    TestLimiter: limiter_test.go:181: Request ratio (0.051948) does not match expected value (0.250000)
    TestLimiter: limiter_test.go:181: Request ratio (0.750025) does not match expected value (0.500000)
    TestLimiter: limiter_test.go:181: Request ratio (0.126047) does not match expected value (0.250000)
    TestLimiter: limiter_test.go:181: Request ratio (0.123928) does not match expected value (0.250000)
```

Reply:
@rjl493456442 I improved the probabilistic testing method a bit because sometimes some values were slightly out of range. Now the test is super stable and quick.
I couldn't reproduce the results seen above though. Your test errors are not just randomly slightly off, they are explicitly wrong. I ran the test several hundred times locally and I haven't seen anything like this. Are you sure you didn't have any code modifications locally? What parameters did you use with `go test`? I tried `cpu=8` but didn't make any difference, my local test results are consistently good now.